### DISCUSSION TITLE
More easy way to switch helm-grep-do-git-grep from default-directory to whole repository.

### DISCUSSION THREAD

**Discussion by zw963** [emacs-helm/helm#2525] [Ideas]
I am using `helm-grep-do-git-grep` for search files in git repo.

Assume, i binding this function to M-r

(global-set-key [(meta r)] 'helm-grep-do-git-grep)

So, when i pressing `M-r`, it will search result in current default-directory.

i consider this is ok as a default, but, if i can't find it, i want to more easy
way to fallback to search on whole repository.

so, maybe, if there is a function, `helm-grep-do-git-grep-whole-repository-within-helm`
i can binding to helm-mode-map, use same keybinding, e.g. `M-r`

so, this is very conveniently.

1. M-r once => search on default-directoy

2. if not found, M-r again, search on whole repository.

If there exits a function like this? Thank you.

**Comment by thierryvolpiatto** [comment]
"Billy.Zheng" ***@***.***> writes:

> I am using helm-grep-do-git-grep for search files in git repo.
>
> Assume, i binding this function to M-r
>
> (global-set-key [(meta r)] 'helm-grep-do-git-grep)
>
> So, when i pressing M-r, it will search result in current default-directory.
>
> i consider this is ok as a default, but, if i can't find it, i want to more easy
> way to fallback to search on whole repository.

C-h k M-r

=>

    [...]
    (helm-grep-do-git-grep ARG)

    Preconfigured `helm' for git-grepping `default-directory'.
    With a prefix arg ARG git-grep the whole repository.

Also you can run helm-find-files to navigate anywhere and launch any
helm grep command from the directory you navigate to.

-- 
Thierry

**Reply by zw963** [reply]
Hi, @thierryvolpiatto , i thought i am not explain clearly what i expected, so, i will create a reproduceable screenshot for explain this, you probably have to see this reply on github directly.

We assume binding key like this:

```cl
(global-set-key (kbd "M-r") 'helm-grep-do-git-grep)
```

1. we assume the cursor on `prog-mode-hook`, when i pressing `M-r`, the output like this:

![image](https://user-images.githubusercontent.com/549126/195978938-bb3711a2-b1ae-48f3-ae12-b104ca3267a2.png)

2.  As you can see, there no meaningful result except the line under cursor, so, the really i want to do for the next expected step is: 

- `search prog-mode-hook in current git project instead` with only one hit.

AFAIK, for now, i have to do is:

(1)  C-g to quit current session.
(2)  Pressing C-u M-r, to search pattern on entire project.

What i expected is, above steps can be combinate into a command, e.g. `helm-quit-and-do-git-grep-on-project`, then binding it in helm-grep-map

In fact, i write a version, please check:

```cl
(defun helm-quit-and-do-git-grep-on-project ()
  "Drop into `helm-grep-do-git-grep' on entire project from `helm'."
  (interactive)
  (with-helm-alive-p
    (helm-run-after-exit #'helm-grep-do-git-grep t)))

(global-set-key (kbd "M-r") 'helm-grep-do-git-grep)
(define-key helm-grep-map (kbd "M-r") 'helm-quit-and-do-git-grep-on-project)
```

This hack is really nice from my point of view.

In fact, we discussed same hack before.

https://github.com/emacs-helm/helm/issues/2437#issuecomment-898991308
https://github.com/emacs-helm/helm/issues/2437#issuecomment-899040211

It's just that the discussion was about search use ag/rg, and when repeat the keybinding, it search on `parent directory` instead of the entire git project.
